crates/bevy_asset/src/io/wasm/opfs.rs

```diff
@@ -0,0 +1,702 @@
+        done: bool,
+    }
+
+    impl Stream for JsStream {
+        type Item = Result<JsValue, JsValue>;
+
+        fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+            if self.done {
+                return Poll::Ready(None);
+            }
+
+            let future = match self.next.as_mut() {
+                Some(val) => val,
+                None => match self.iter.next().map(JsFuture::from) {
+                    Ok(val) => {
+                        self.next = Some(val);
+                        self.next.as_mut().unwrap()
+                    }
+                    Err(e) => {
+                        self.done = true;
+                        return Poll::Ready(Some(Err(e)));
+                    }
+                },
+            };
+
+            match Pin::new(future).poll(cx) {
+                Poll::Ready(res) => match res {
+                    Ok(iter_next) => {
+                        let next = iter_next.unchecked_into::<IteratorNext>();
+                        if next.done() {
+                            self.done = true;
+                            Poll::Ready(None)
+                        } else {
+                            self.next.take();
+                            Poll::Ready(Some(Ok(next.value())))
+                        }
+                    }
+                    Err(e) => {
+                        self.done = true;
+                        Poll::Ready(Some(Err(e)))
+                    }
+                },
+                Poll::Pending => Poll::Pending,
+            }
+        }
+    }
+
+    impl From<AsyncIterator> for JsStream {
+        fn from(value: AsyncIterator) -> Self {
+            Self {
+                iter: value,
+                next: None,
+                done: false,
+            }
+        }
+    }
+
+    /// Extension trait providing access to the async iterator methods on [`FileSystemDirectoryHandle`]
+    /// which are currently missing from [`wasm-bindgen`](`wasm_bindgen`)
+    pub(crate) trait FileSystemDirectoryHandleExt {
+        fn keys(&self) -> AsyncIterator;
+    }
+
+    impl FileSystemDirectoryHandleExt for FileSystemDirectoryHandle {
+        /// The `keys()` method.
+        ///
+        /// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/keys)
+        ///
+        /// *This API requires the following crate features to be activated: `FileSystemDirectoryHandle`*
+        fn keys(&self) -> AsyncIterator {
+            #[wasm_bindgen(
+                inline_js = "export function get_keys_for_handle(a) { return a.keys(); }"
+            )]
+            extern "C" {
+                fn get_keys_for_handle(a: &FileSystemDirectoryHandle) -> AsyncIterator;
+            }
+
+            get_keys_for_handle(self)
+        }
+    }
+
+    /// Uses channels to create a [`Send`] + [`Sync`] wrapper around a [`Stream`].
+    pub(crate) struct IndirectStream<T> {
```

Comment:
Unfortunately `boxed` won't work in this case, since it requires the inner stream be `Send`. The issue we have is the `read_directory` method requires a stream which implements `Stream<Item=PathBuf> + Send + Unpin`, but I can only produce a `Stream<Item=PathBuf> + Unpin` because the inner [JS handles are explicitly `!Send`](https://github.com/rustwasm/wasm-bindgen/issues/2833#issuecomment-1063055053).

As such, the workaround I use here is to use channels to leave the `!Send` resources in-place (thrown onto the JS task queue using `spawn_local`), letting us send a dummy `Stream` object that just wraps the channel.

I find this _particularly_ frustrating as most of futures in WASM is `!Send` already, so I'm half-tempted to make `PathStream` `ConditionalSend` as a workaround, but I want to avoid changing any existing APIs as much as possible.